Allow ordering of channels

njhanley opened the pull request:
Fixes #104

psychon commented:
Could you briefly outline what this change does? From a quick look, this seems only related to the config, but it does not actually do anything on the IRC wire protocol.

So, how exactly do you understand "order of channels" and how is this order communicated to IRC clients? What happens if channel number two cannot be joined (e.g. because one is banned)? Would this mean that channel number three cannot be attempted to be joined since that would mess up the order that the client sees?

Basically, I always understood #104 as something that is just not supported by the IRC protocol and so I am curious to understand how you solved these issues, but too lazy to read the patch carefully to understand this.

njhanley commented:
The first commit makes CConfig preserve the order of subconfigs as they appear in znc.conf. The second adds commands to manipulate the order of channels in a network's channel vector.

> From a quick look, this seems only related to the config, but it does not actually do anything on the IRC wire protocol.

No changes to the wire protocol are needed. ZNC already joins clients to channels in a defined order, specifically lexicographically by name. These changes put that order under the user's control.

> So, how exactly do you understand "order of channels" and how is this order communicated to IRC clients?

The order of channels in the context of #104 is the order in which ZNC will attach a client to each channel. There is nothing new to communicate to clients.

> What happens if channel number two cannot be joined (e.g. because one is banned)?

The same behavior as current; the channel will be skipped. It will still occupy a place in the network's channel vector and can be ordered with the two new commands.

> Basically, I always understood #104 as something that is just not supported by the IRC protocol and so I am curious to understand how you solved these issues, but too lazy to read the patch carefully to understand this.

I'm not sure what you mean by unsupported by the IRC protocol.

psychon commented:
So, if znc fails to join a channel for whatever reason and the user then later joins that channel "by hand", the resulting order in the irc client will be "wrong". Okay.

Still, doing what you did might be enough for some people. Thanks.

DarthGandalf commented on test/ConfigTest.cpp:
```diff
@@ -146,6 +145,10 @@ TEST_F(CConfigSuccessTest, SubConf8) {
                  "->a/B\nfoo=bar\nfooo=bar\n<-\n");
 }
+TEST_F(CConfigSuccessTest, SubConf9) {
+    TEST_SUCCESS("<foo b>\n</foo>\n<foo a>\n</foo>",
```
Please add comment that it's supposed to test not sorting the subconfs by name

DarthGandalf commented on src/Config.cpp:
```diff
@@ -122,14 +123,17 @@ bool CConfig::Parse(CFile& file, CString& sErrorMsg) {
+                auto& nameset = pActiveConfig->m_SubConfigNameSets[sTagLower];
 
-                if (it != conf.end())
+                const auto& it = nameset.find(sName);
```
auto

Reference is not appropriate here, even though technically it works due to lifetime extension

DarthGandalf commented on modules/data/webadmin/tmpl/add_edit_network.tmpl:
```diff
@@ -197,7 +197,7 @@
 					</thead>
 
 					<tbody>
-			<? LOOP ChannelLoop SORTASC=Name ?>
```
This belongs to separate commit

DarthGandalf commented on src/ClientCommand.cpp:
```diff
@@ -438,6 +438,50 @@ void CClient::UserCommand(CString& sLine) {
         }
+    } else if (sCommand.Equals("MOVECHAN")) {
+        if (!m_pNetwork) {
+            PutStatus("You must be connected with a network to use this command");
```
Missing `t_s()`, and below

I wonder how I missed it some other strings in this file

DarthGandalf commented on src/ClientCommand.cpp:
```diff
@@ -438,6 +438,50 @@ void CClient::UserCommand(CString& sLine) {
+        // use one-based indexing
+        unsigned int uIndex = sTarget.ToUInt();
+        if (uIndex == 0 || uIndex > m_pNetwork->GetChans().size()) {
+            PutStatus(t_f("Invalid index {1}")(uIndex));
```
It's not obvious what number user should pass to the command. Probably ListChans should show the current index?

DarthGandalf commented on src/IRCNetwork.cpp:
```diff
@@ -942,6 +942,38 @@ bool CIRCNetwork::DelChan(const CString& sName) {
+    if (i == m_vChans.size()) return false;
+
+    const auto pChan = m_vChans[i];
+    for (int delta = i < uIndex ? 1 : -1; i != uIndex; i += delta)
```
erase() followed by insert() would be much simpler

DarthGandalf commented on src/IRCNetwork.cpp:
```diff
@@ -942,6 +942,38 @@ bool CIRCNetwork::DelChan(const CString& sName) {
+        if (m_vChans[j]->GetName().Equals(sChan2)) break;
+    if (j == m_vChans.size()) return false;
+
+    const auto pChan = m_vChans[i];
```
std::swap?

njhanley commented:
I can't find any tests for the existing commands. Where might I add the new ones?

Yes, reordering channels in the webadmin would be useful. I'll see what I can do.

DarthGandalf commented:
> I can't find any tests for the existing commands. Where might I add the new ones?

It's in https://github.com/znc/znc/tree/master/test/integration/tests

DarthGandalf commented on src/IRCNetwork.cpp:
```diff
@@ -942,6 +942,35 @@ bool CIRCNetwork::DelChan(const CString& sName) {
     return false;
 }
 
+bool CIRCNetwork::MoveChan(const CString& sName, unsigned int uIndex) {
```
Every time this function is called, you're checking first whether the channel exists.

Maybe pass the `CChannel*` instead, to avoid looking it up again? Same for swap

Another option is to return a better error message from here instead of just "false", then you won't need to check it before calling, but just need to show the error message returned from here

njhanley replied:
Passing the `CChan*` would be cleaner. This and removing the redundant check on `uIndex` would make both `MoveChan` and `SwapChan` void functions and leave all input validation at the call sites.

DarthGandalf replied:
I've just realized that `CChan*` wouldn't be enough, because it will still require searching for it in the vector.

There are other functions here already which return the error via `CString& sError`

njhanley commented on src/IRCNetwork.cpp:
```diff
@@ -942,6 +942,35 @@ bool CIRCNetwork::DelChan(const CString& sName) {
+    if (uIndex >= m_vChans.size()) return false;
+
+    auto it = m_vChans.begin();
+    for (; it != m_vChans.end(); ++it)
```
Initially I thought this would require `std::find_if` with a rather ugly lambda. I see now that passing a `CChan*` from `CIRCNetwork::FindChan` would work.

njhanley replied:
I don't see how this would be an improvement. Also `std::find_if` would be needed.

njhanley commented on modules/data/webadmin/tmpl/add_edit_network.tmpl:
```diff
@@ -213,6 +213,15 @@
 					</tbody>
 				</table>
+				<script>
+				$("#channels").sortable({
```
That's correct. I will implement a non-JS interface.

njhanley replied:
I'd like to rework the channels interface by copying the design of the network servers list.

Instead of the current table, the non-JS UI would be a textarea where each line specifies the configuration of a channel. With JS, the textarea would be replaced by a table with more user friendly inputs and draggable rows. I would also remove the add/edit channel page and move its options to the new interface.

Alternatively I could add a textarea (hidden with JS) to specify the channels in the desired order. This approach would not change the existing UI.

I believe the first approach is cleaner. What are your thoughts?

DarthGandalf replied:
> I'd like to rework the channels interface by copying the design of the network servers list.

Tbh, I'm not entirely happy with what I did there. Currently the JS editor edits the (hidden) text area, while it would probably be cleaner to send a flag to the server whether to take input from the text area or from separate fields resulting from JS editor directly. Not sure.

> the non-JS UI would be a textarea where each line specifies the configuration of a channel

The channel config is more complicated than list of servers, so this would make webadmin essentially unusable for editing channels without JS. Maybe you have a mock of the new UI if I misunderstood what do you mean?

> I would also remove the add/edit channel page and move its options to the new interface.

This will break modules which add parts of UI to that page, e.g. stickychan. How would you present that in a single text field?

njhanley replied:
> Maybe you have a mock of the new UI if I misunderstood what do you mean?

Something like this.

![channel serialization mock](https://i.imgur.com/fQGzHCm.png)

Here's a possible grammar in EBNF.
```ebnf
channel = name , space ,
          [ "key:" , key , space ] ,
          buffer size , space ,
          auto clear buffer , space ,
          save , space ,
          enabled , space ,
          detached , space ,
          default modes , newline ;
name = ? irc channel name ? ;
key = ? irc channel key ? ;
buffer size = "default" | integer ;
auto clear buffer = "default" | boolean ;
save = boolean ;
enabled = boolean ;
detached = boolean ;
default modes = ? irc channel modes, may contain spaces ? ;

space = " " , { " " } ;
newline = "\n" ;
boolean = "false" | "true" ;
integer = "0" | "1"..."9" , { "0"..."9" } ;
```

> This will break modules which add parts of UI to that page, e.g. stickychan.

I overlooked that section of the page, my bad. I'll abandon this idea.

njhanley commented:
I've added `sError` parameters to MoveChan and SwapChans and removed the redundant checks/searches.
Also I wrote a different interface for rearranging channels in the webadmin. It works without JS.

![channel indexes mock](https://i.imgur.com/dHj9d5e.png)

DarthGandalf commented:
Thanks!
I've added similar CSS to other skins to not make the index field so wide
